sapicore/engine/component/__init__.py

```diff
@@ -0,0 +1,183 @@
+
+    configuration: dict, optional
+        Configuration dictionary used during parameter initialization.
+
+    device: str, optional
+        Specifies a hardware device on which to process this object's tensors, e.g. "cuda:0".
+        Must be a valid torch.device() string value. Defaults to "cpu".
+
+    kwargs:
+        Additional instance attributes that the user may set.
+
+    See Also
+    --------
+    :class:`~tree_config.Configurable`
+    :class:`~utils.loggable.Loggable`
+    :class:`~utils.sweep.Sweep`
+
+    """
+
+    def __init__(self, identifier: str = None, configuration: dict = None, device: str = "cpu", **kwargs):
+        super().__init__()
+
+        # model-related common instance attributes.
+        self.identifier = identifier
+        self.configuration = configuration
+
+        # hardware device on which to process this component.
+        self.device = device
+
+        # simulation-related tracking variables.
+        self.simulation_step = 0
+        self.dt = DT
+
+        # initialize configurable and loggable attributes with dummy tensors.
+        for prop in self._config_props_:
+            setattr(self, prop, torch.zeros(1, dtype=torch.float, device=self.device))
+
+        for prop in self._loggable_props_:
+            self.register_buffer(prop, torch.zeros(1, dtype=torch.float, device=self.device))
+
+        # developer may override or define arbitrary attributes at instantiation.
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def configure(self, configuration: dict[str, Any], log_destination: str = None):
+        """Applies a configuration to this object by adding the keys of `configuration` as instance attributes,
+        initializing their values, and updating the `_config_props_` tuple to reflect the new keys.
+
+        Also updates `_loggable_props_` for this object if the configuration includes the key "loggable".
+
+        Parameters
+        ----------
+        configuration: dict or str
+            Parameters to be applied to this neuron, given as a dictionary.
+
+        log_destination: str, optional
+            Path to a destination YAML file where the configuration will be saved.
+            If not provided, the configuration is not saved to file.
+
+        Note
+        ----
+        This method is generic and can be used without alterations by all derivative classes.
+
+        The sub-dictionary `model` may include the special `sweep` key, which specifies settings for the
+        heterogeneous initialization of certain parameters in the form of a dictionary.
+
+        """
+        # apply this object's configuration by adding or updating instance variables.
+        apply_config(self, configuration)
+
+        # override _loggable_properties_ if provided in the configuration dictionary.
+        if configuration.get("loggable"):
+            self._loggable_props_ = configuration.get("loggable")
```

Comment:
I'll do that for this PR, but I'd strongly recommend we take a few days to clean up the loggable API and overhaul the docstrings. It's very opaque in its current state:

* Why have a `_loggable_props` cache in addition to `_loggable_props_`? 
* `_loggable_props` isn't included in the `Loggable` docstring, so users have to guess or dig into the code to figure it out. The word "cache" doesn't appear anywhere.
* Why is one a tuple and the other a list? 
* Why is there a standalone `get_loggable_properties` but no `set_loggable_properties`? 
* Why is `load_save_get` one consolidated function and not three separate ones? 
* Why were all of these functions implemented outside `Loggable` and not as methods?

Users shouldn't be expected to invest time figuring out the ins and outs of this module--it should be clear how to leverage it to create, read, update, or delete loggable properties across a class hierarchy. As I understand it (maybe naively), each of these four operations should be supported separately for files and objects.

Reply:
I can work on the docstrings, but about the opaqueness, end users (who set the properties at the class level and get the list) weren't meant to look into those details (i.e. cache) so you are more in the framework dev category for this hence why you need to know about these internal details especially because you're setting them dynamically - which I didn't wrote docs for. But I can fix that. Because most users shouldn't have to know about cache etc and should only use the API (i.e. reading `self.loggable_props` and setting `_loggable_props_` at def time).

* More specifically, the reason why there is a cache in loggable is because the exact loggable props may be looked up multiple times while the network is running and whether something is loggable my change during network running, hence cache is needed. Otherwise the computational cost of looking it up every time is too much. Configurable has something identical, but there it's actually named `_config_props_cache`.
* The tuple (`_config_props_`/`_loggable_props_`) is exposed to the end use and it's a tuple meant to indicate it's immutable and should not be changed other than at the class level when users annotate their classes. I'm using a list for cache as lists are more convenient to work with, are faster, and are my go to structure vs tuple. Similarly, users are returned a list of config props/loggables `with self.loggable_props` and `self._config_props` because I prefer to provide lists where possible rather than tuples because they are more convenient.
* There's no `set_loggable_properties` because I'm not sure what that would do? `get_loggable_properties` gives you all the properties to be logged given an object and a config dict that indicates which props are to be logged. I'm not sure what a setter would do?
* about `load_save_get_loggable_properties`, it is a shorthand for the three functions `dump_loggable`, `read_loggable_from_file`, `dump_loggable`, and `get_loggable_properties` so they are separateable.
* The reason they are external methods is that I didn't want to pollute the object namesapce with these methods. Because these methods are really only meant for us framework devs and those creating loggers etc. So I think it's better for them to be functions that accept object rather than class methods. But I can be convinced otherwise.

Reply:
Makes sense. I think all comments related to this can be resolved once the `_loggable_props` cache is fixed to update the moment `_loggable_props_` is set (as opposed to after the first time a loggable is accessed; see below). Also, now that child class cfg_props are no longer initialized at `Component`, I can use `_config_props` everywhere.

If log/cfg props can be updated during a run (or even during initialization), shouldn't the API include setters? Again, naively, something like `get_loggables(obj | Path)`, `get_configurables(obj | Path)`, `set_loggables(obj, tuple | list, dump: bool)`, `set_configurables(obj, tuple | list, dump: bool)` would be more intuitive and cover all the bases. Come to think of it, there's a lot of duplication (functionality-wise) between Loggable and Configurable--why not have a parent `Property` class that would handle loading/dumping from/to YAML, maintain caches, etc.? At the end of the day they're both a tuple[str] with a list[str] cache, endowed with methods that know how to accumulate them across a class hierarchy.